File: ATSP/src/Runners/Execution.cs
using System.Collections.Generic;

namespace ATSP.Runners
{
    public class Execution
    {
        public override bool Equals(object obj)
        {
            if(obj is Execution execution)
            {
                return execution.AlgorithmName == this.AlgorithmName &&
                    execution.ExecutionID == this.ExecutionID;
            }
            return false;
        }

        public override int GetHashCode()
        {
            return AlgorithmName.GetHashCode() + this.ExecutionID.GetHashCode();
        }

        public string AlgorithmName { get; set; }
        public ulong ExecutionID { get; set; }
        public double Time { get; set; }
        public uint Steps { get; set; }
        public uint Cost { get; set; }
        public uint BestKnownCost { get; set;}
        public uint NumberOfImprovements { get; set; }
        public List<uint> IntermediateCosts { get; set; }
        public double SimilarityWithBest { get; set; }
        public uint InitialCost { get; set; }

        public double Quality => (double)(Cost - BestKnownCost) / BestKnownCost ;

        // it is easier to calculate effectiveness after importing data to python
        // effectiveness = 
        //  (Number_of_improvements/maxImprovements) / (Execution_time / meanExecutionTime * Execution_steps / maxSteps * Quality)
        // public double CalculateEffectiveness(uint worstResult)
        //     => Effectiveness = 1 - (double)(Cost - BestKnownCost) / (worstResult - BestKnownCost);


        public double InitialToFinalExecutionRatio
            => (double)(InitialCost) / Cost;

        public double Effectiveness { get; private set; }

        public uint[] FinalSolution { get; set; }
    }
}

File: Raport/execution.py
import numpy as np

class Execution:
    def __init__(self):
        self.time = 0.0
        self.steps = 0
        self.cost = 0
        self.intermediate_costs = []
        self.best_known_cost = 0
        self.initial_cost = 0
        self.similarity = 0.0
        self.number_of_improvements = 0
        self.solution = []

    @property
    def initial_to_final_execution_improvement(self):
        if self.cost == 0:
            return 0
        return (float)(self.initial_cost) / self.cost

    @property
    def quality(self):
        return (float)(self.cost - self.best_known_cost) / self.best_known_cost

    @property
    def simple_quality(self):
        return (self.cost - self.best_known_cost)

    def get_effectiveness(self, worst_result):
        distance = worst_result - self.best_known_cost
        return 1 - float(self.cost - self.best_known_cost) / distance

    def get_effectiveness2(self, worst_result):
        return self.get_effectiveness(worst_result) * (1/self.time)

    def __len__(self):
        return len(self.intermediate_costs)

    def first_best_quality(self):
        if len(self.intermediate_costs)>0:
            return (self.intermediate_costs[0]- self.best_known_cost) / self.best_known_cost, self.quality
        return 0,0
